### PR TITLE
[cxxmodules] Fix ROOT-11002 -- Vc plus GMI

### PR DESCRIPTION
**Pull request by vgvassilev**
[cxxmodules] Do not perform version checks for Vc.

We do not link against libVc.a thus it makes no sense to check for version compatibility between libraries and header files. This fixes a bug where upon building the modules.idx we run the static ctor runLibraryAbiCheck which fails to find the corresponding symbol.

### CODE REVIEW

**Comment by vgvassilev** [comment]
@lmoneta, shouldn't we link the whole archive of libVc.a into MathCore? We see some failures such as:
```root.exe -q -l -b test/testGenVectorVc.cxx 

Processing test/testGenVectorVc.cxx...
IncrementalExecutor::executeFunction: symbol '_ZN4Vc_13SSE9c_general14signMaskDoubleE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
You are probably missing the definition of Vc_1::SSE::c_general::signMaskDouble
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZN4Vc_13SSE9c_general13absMaskDoubleE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
You are probably missing the definition of Vc_1::SSE::c_general::absMaskDouble
Maybe you need to load the corresponding shared library?
IncrementalExecutor::executeFunction: symbol '_ZN4Vc_16Common10AllBitsSetE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
You are probably missing the definition of Vc_1::Common::AllBitsSet
Maybe you need to load the corresponding shared library?
warning: Failed to call `testGenVectorVc()` to execute the macro.
Add this function or rename the macro. Falling back to `.L`.```


Where: 
``` nm externals/usr/local/lib/libVc.a | grep _ZN4Vc_13SSE9c_general14signMaskDoubleE
0000000000000580 R _ZN4Vc_13SSE9c_general14signMaskDoubleE
```

File: math/mathcore/CMakeLists.txt [hunk 1]
```diff
@@ -106,8 +106,12 @@ if(runtime_cxxmodules)
   # FIXME: Revise after a llvm upgrade or reproduce it outside rootcling.
   list(REMOVE_ITEM HEADERS "Math/Math.h")
 
-  if(veccore)
-    set(dictoptions "-mByproduct" "Vc")
+  if(veccore OR vc)
+    # We do not link against libVc.a thus it makes no sense to check for
+    # version compatibility between libraries and header files. This fixes
+    # ROOT-11002 where upon building the modules.idx we run the static ctor
```
Comment (Axel-Naumann) [inline math/mathcore/CMakeLists.txt:112]: Why does building the module index *run* anything?

Comment (vgvassilev) [reply math/mathcore/CMakeLists.txt:112]: Because when building the modules.idx we need to load all files. That also mean doing static initialization for the objects as the usual rules apply.

**Comment by vgvassilev** [comment]
> @lmoneta, shouldn't we link the whole archive of libVc.a into MathCore? We see some failures such as:
> 
> ```
> root.exe -q -l -b test/testGenVectorVc.cxx 
> 
> Processing test/testGenVectorVc.cxx...
> IncrementalExecutor::executeFunction: symbol '_ZN4Vc_13SSE9c_general14signMaskDoubleE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
> You are probably missing the definition of Vc_1::SSE::c_general::signMaskDouble
> Maybe you need to load the corresponding shared library?
> IncrementalExecutor::executeFunction: symbol '_ZN4Vc_13SSE9c_general13absMaskDoubleE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
> You are probably missing the definition of Vc_1::SSE::c_general::absMaskDouble
> Maybe you need to load the corresponding shared library?
> IncrementalExecutor::executeFunction: symbol '_ZN4Vc_16Common10AllBitsSetE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
> You are probably missing the definition of Vc_1::Common::AllBitsSet
> Maybe you need to load the corresponding shared library?
> warning: Failed to call `testGenVectorVc()` to execute the macro.
> Add this function or rename the macro. Falling back to `.L`.
> ```
> 
> Where:
> 
> ```
> nm externals/usr/local/lib/libVc.a | grep _ZN4Vc_13SSE9c_general14signMaskDoubleE
> 0000000000000580 R _ZN4Vc_13SSE9c_general14signMaskDoubleE
> ```

@lmoneta ping.

**Comment by vgvassilev** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
> > @lmoneta, shouldn't we link the whole archive of libVc.a into MathCore? We see some failures such as:
> > ```
> > root.exe -q -l -b test/testGenVectorVc.cxx 
> > 
> > Processing test/testGenVectorVc.cxx...
> > IncrementalExecutor::executeFunction: symbol '_ZN4Vc_13SSE9c_general14signMaskDoubleE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
> > You are probably missing the definition of Vc_1::SSE::c_general::signMaskDouble
> > Maybe you need to load the corresponding shared library?
> > IncrementalExecutor::executeFunction: symbol '_ZN4Vc_13SSE9c_general13absMaskDoubleE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
> > You are probably missing the definition of Vc_1::SSE::c_general::absMaskDouble
> > Maybe you need to load the corresponding shared library?
> > IncrementalExecutor::executeFunction: symbol '_ZN4Vc_16Common10AllBitsSetE' unresolved while linking function '_GLOBAL__sub_I_cling_module_9'!
> > You are probably missing the definition of Vc_1::Common::AllBitsSet
> > Maybe you need to load the corresponding shared library?
> > warning: Failed to call `testGenVectorVc()` to execute the macro.
> > Add this function or rename the macro. Falling back to `.L`.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Where:
> > ```
> > nm externals/usr/local/lib/libVc.a | grep _ZN4Vc_13SSE9c_general14signMaskDoubleE
> > 0000000000000580 R _ZN4Vc_13SSE9c_general14signMaskDoubleE
> > ```
> 
> @lmoneta ping.

@lmoneta, ping 2.

**Comment by lmoneta** [comment]
@vgvassilev , I don't understand this problem.
Mathcore should not depend on Vc, we should not add Vc as a static library.
There is maybe some dependency only in the test programs, but in that case we can link then Vc statically when building the tests

**Comment by vgvassilev** [comment]
So if that’s the case then this PR is good to go.

File: math/mathcore/CMakeLists.txt (hunk 1 above)
Comment (Axel-Naumann) [reply math/mathcore/CMakeLists.txt:112]: rootcling usually has syntax-only mode, i.e. no execution engine. What am I missing?

Comment (Axel-Naumann) [reply math/mathcore/CMakeLists.txt:112]: @vgvassilev ping

Comment (vgvassilev) [reply math/mathcore/CMakeLists.txt:112]: I do not understand your question. Rootcling generates a module from headers which have static initializers. When loading that module by root we run the initializers. This means that the way we build the .so file and the module needs to be the same. We do not find the Vc versioning symbol anywhere (it is not in MathCore) meaning we need to just build the module in the same way.

Comment (Axel-Naumann) [reply math/mathcore/CMakeLists.txt:112]: My understanding was that the error message is emitted by `rootcling` (likely from the comment "upon building the modules.idx we run the static ctor"), not `root`. Your latest comment here clarifies it, but to avoid my confusion maybe reword:
```suggestion
    # ROOT-11002 where modules.idx triggers running the Vc's static ctor
```

Comment (vgvassilev) [reply math/mathcore/CMakeLists.txt:112]: The modules.idx is built byte root.exe.

**Review by bellenot — APPROVED** [review]
LGTM

**Review by Axel-Naumann — APPROVED** [review]
